src/asmcnc/job/job_data.py

```diff
@@ -389,3 +400,209 @@ def post_job_data_update_post_send(self):
+
+            self.job_recovery_filepath = job_recovery_info[0]
+            self.job_recovery_cancel_line = int(job_recovery_info[1])
+
+        except:
+            print("Could not read recovery info")
+            print(str(traceback.format_exc()))
+
+    def write_to_recovery_file(self, cancel_line):
+        try:
+            # Account for number of lines added in by the software when running file
+            cancel_line -= self.job_recovery_offset
+            self.job_recovery_offset = 0
+
+            with open(self.job_recovery_info_filepath, 'w+') as job_recovery_info_file:
+                job_recovery_info_file.write(self.filename + "\n" + str(cancel_line))
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = self.filename
+            self.job_recovery_cancel_line = cancel_line
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not write recovery info")
+            print(str(traceback.format_exc()))
+
+    def clear_recovery_file(self):
+        try:
+            open(self.job_recovery_info_filepath, 'w').close()
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = ''
+            self.job_recovery_cancel_line = 0
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not clear recovery info")
+            print(str(traceback.format_exc()))
+
+
+    def reset_recovery(self):
+        self.job_recovery_selected_line = -1
+        self.job_recovery_gcode = []
+        self.job_recovery_offset = 0
+
+    def generate_recovery_gcode(self):
+        try:
+            recovery_gcode = []
+
+            # Recover modal gcodes
+
+            # Coordinate System Select
+            coord_system_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G5[4-9]", s)), None)
+            if coord_system_line:
+                recovery_gcode.append(re.search("G5[4-9](\.[1-3])?", coord_system_line).group(0))
+
+            # Plane selection
+            plane_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G1[7-9]", s)), None)
+            if plane_line:
+                recovery_gcode.append(re.search("G1[7-9]", plane_line).group(0))
+
+            # Distance mode
+            distance_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]([A-Z]|\s|$)", s)), None)
+            if distance_line:
+                # Recovery not allowed if G91 is used
+                if re.search("G91([A-Z]|\s|$)", distance_line):
+                    return False, 'This job contains the G91 g-code, so it cannot be recovered.'
```

Comment:
Change error message to: 

"The last positioning declaration was incremental (G91), and therefore this job cannot be recovered."